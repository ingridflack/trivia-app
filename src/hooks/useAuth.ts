import { useState } from "react";

interface UserData {
  name: string;
  avatar?: string;
  username: string;
  id: string;
}

interface UserData {
  token: string;
  user: UserData;
}

const useAuth = () => {
  const [userData, setUserData] = useState<UserData | null>(() => {
    const storedUserData = localStorage.getItem("userData");
    return storedUserData ? JSON.parse(storedUserData) : null;
  });

  const [token, setToken] = useState<string | null>(() => {
    return localStorage.getItem("token") || null;
  });

  const saveUserData = (data: UserData) => {
    localStorage.setItem("token", data.token);
    localStorage.setItem("userData", JSON.stringify(data.user));
    setToken(data.token);
    setUserData(data.user);
  };

  const clearUserData = () => {
    localStorage.removeItem("token");
    localStorage.removeItem("userData");
    setToken(null);
    setUserData(null);
  };

  return { userData, token, saveUserData, clearUserData };
};

export default useAuth;
