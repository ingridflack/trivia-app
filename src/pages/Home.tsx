import { Box, Button, Container, Image, Text } from "@chakra-ui/react";
import Header from "../components/Header";
import { Link } from "react-router-dom";

export default function Home() {
  return (
    <>
      <Header />
      <Box width="100%" backgroundColor="#AE79F8" overflow="hidden">
        <Container paddingY="80px" position="relative">
          <Text
            fontSize="6xl"
            lineHeight="1.2"
            fontWeight="bold"
            color="white"
            maxWidth="550"
          >
            Challenge yourself in a trivia match!
          </Text>

          <Button
            backgroundColor="green.400"
            color="white"
            width="180px"
            _hover="green.400"
            marginTop="38px"
            as={Link}
            to="/trivia/config"
          >
            Start a trivia
          </Button>
          <Image
            src="/src/assets/home.svg"
            alt="Home image"
            position="absolute"
            right="-160px"
            bottom="-60px"
            width="674px"
            height="418px"
            pointerEvents="none"
          />
        </Container>
      </Box>
    </>
  );
}
