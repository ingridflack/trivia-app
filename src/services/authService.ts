import axios from "axios";
import { BASE_URL } from "../config/baseUrl";

interface LoginParams {
  email: string;
  password: string;
}

export const login = ({ email, password }: LoginParams) =>
  axios.post(`${BASE_URL}/login`, { email, password });
