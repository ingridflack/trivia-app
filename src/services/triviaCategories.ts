import axios from "../config/api";
import { Category, ApiResponse } from "../types/sharedTypes";

interface CategoryResponse extends ApiResponse {
  categories: Category[];
}

export const getCategories = () =>
  axios.get<CategoryResponse>("/trivia/categories");
