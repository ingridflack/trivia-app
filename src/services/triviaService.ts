import axios from "../config/api";
import { Category, ApiResponse, Trivia } from "../types/sharedTypes";

interface CategoryResponse extends ApiResponse {
  categories: Category[];
}

interface TriviaQueryParams {
  category: number;
  difficulty: string;
  amount: number;
  type?: string;
}

interface TriviaResponse extends ApiResponse {
  triviaId: string;
}

interface TriviaByIdResponse extends ApiResponse {
  trivia: Trivia;
}

export const getCategories = () =>
  axios.get<CategoryResponse>("/trivia/categories");

export const createTrivia = (params: TriviaQueryParams) =>
  axios.post<TriviaResponse>(
    `/trivia?amount=${params.amount}&type=&difficulty=${params.difficulty}&category=${params.category}`
  );

export const getTriviaById = (triviaId: string) =>
  axios.get<TriviaByIdResponse>(`/trivia/${triviaId}`);
