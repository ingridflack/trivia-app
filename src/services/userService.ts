import axios from "../config/api";
import { UserSearch } from "../types/sharedTypes";

export const search = (params: Partial<UserSearch>) =>
  axios.get<UserSearch[]>("/users/search", { params });

export const getUser = () => axios.get("/users");
