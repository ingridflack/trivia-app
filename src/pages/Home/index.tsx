import {
  Box,
  Button,
  Container,
  Heading,
  Image,
  Skeleton,
  Stack,
  Text,
} from "@chakra-ui/react";
import Header from "../../components/Header";
import { Link, useNavigate } from "react-router-dom";
import { PendingTriviaList } from "../../components/PendingTriviaList";
import { useEffect, useState } from "react";
import * as TriviaService from "../../services/triviaService";
import { CreateTriviaBody, PendingTrivia } from "../../types/sharedTypes";
import { Footer } from "../../components/Footer";
import useAuth from "../../hooks/useAuth";
import { TRIVIA_CATEGORIES } from "../../constants/trivia";
import { CategoryCard } from "../../components/CategoryCard";

export default function Home() {
  const [pendingTrivia, setPendingTrivia] = useState<PendingTrivia[]>([]);
  const [isLoading, setIsLoading] = useState(true);
  const { userData } = useAuth();
  const navigate = useNavigate();

  useEffect(() => {
    const fetchPendingTrivia = async () => {
      try {
        const { data } = await TriviaService.getPendingTrivia();
        setPendingTrivia(data.pendingTrivia);
      } catch (error) {
        console.log(error);
      } finally {
        setIsLoading(false);
      }
    };

    fetchPendingTrivia();
  }, []);

  const handleCategoryCardClick = async (params: CreateTriviaBody) => {
    try {
      if (!userData) {
        localStorage.setItem("triviaConfig", JSON.stringify(params));
        navigate("/login");
        return;
      }

      const { data } = await TriviaService.createTrivia(params);

      navigate(`/trivia/${data.triviaId}`);
    } catch (error) {
      console.log(error);
    }
  };

  return (
    <>
      <Header />

      <Box as="main">
        <Box width="100%" backgroundColor="purple.500" overflow="hidden">
          <Container
            padding={{ base: "40px 20px", md: "80px 0" }}
            position="relative"
          >
            <Text
              fontSize={{ base: "3xl", md: "6xl" }}
              lineHeight="1.2"
              fontWeight="bold"
              color="white"
              maxWidth="550"
            >
              Challenge yourself in a trivia match!
            </Text>

            <Button
              backgroundColor="green.400"
              color="white"
              width="180px"
              _hover="green.400"
              marginTop="38px"
              as={Link}
              to="/trivia/config"
            >
              Start a trivia
            </Button>
            <Image
              src="/assets/home.svg"
              alt="Home image"
              position="absolute"
              right="-160px"
              bottom="-60px"
              width="674px"
              height="418px"
              pointerEvents="none"
              display={{ base: "none", md: "none", lg: "block" }}
            />
          </Container>
        </Box>

        {userData && (
          <Container padding={{ base: "40px 20px", md: "80px 0 120px" }}>
            <Box as="header" marginBottom={{ base: "20px", md: "40px" }}>
              <Heading fontSize="3xl">Pending trivias</Heading>
            </Box>

            {isLoading ? (
              <Stack
                display="grid"
                gap="16px"
                gridTemplateColumns={{
                  base: "repeat(1fr)",
                  md: "repeat(4, 1fr)",
                }}
              >
                <Skeleton height="138px" width="100%" borderRadius="md" />
                <Skeleton height="138px" width="100%" borderRadius="md" />
                <Skeleton height="138px" width="100%" borderRadius="md" />
                <Skeleton height="138px" width="100%" borderRadius="md" />
              </Stack>
            ) : pendingTrivia.length ? (
              <PendingTriviaList pendingTriviaList={pendingTrivia} />
            ) : (
              <Text>No pending trivias</Text>
            )}
          </Container>
        )}

        <Box
          as="section"
          padding={{ base: "40px 20px", md: "80px 0 120px" }}
          backgroundColor="gray.100"
          boxShadow="0px 30px 100px 0 rgba(39, 39, 39, 0.2);"
        >
          <Box
            as="header"
            textAlign="center"
            marginBottom={{ base: "20px", md: "60px" }}
          >
            <Heading fontSize={{ base: "3xl", md: "5xl" }} marginBottom="20px">
              Show off your knowledge
            </Heading>
            <Text fontSize={{ base: "md", md: "xl" }}>
              Take a quiz and impress your friends with what you know!
            </Text>
          </Box>

          <Container
            display="grid"
            gridTemplateColumns={{ base: "repeat(1fr)", md: "repeat(5, 1fr)" }}
            gap="16px"
            padding={{ base: "0", md: "0 0" }}
          >
            {Object.values(TRIVIA_CATEGORIES).map((category) => (
              <CategoryCard
                key={category.title}
                title={category.title}
                icon={category.icon}
                color={category.color}
                onCardClick={() => handleCategoryCardClick(category.params)}
              />
            ))}
          </Container>
        </Box>
      </Box>

      <Footer />
    </>
  );
}
