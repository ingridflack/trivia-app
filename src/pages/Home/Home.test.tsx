import { cleanup, render, waitFor } from "@testing-library/react";
import { afterEach, describe, expect, it, Mock, vi } from "vitest";
import useAuth from "../../hooks/useAuth";
import { MemoryRouter } from "react-router-dom";
import Home from ".";
import * as TriviaService from "../../services/triviaService";

vi.mock("../../hooks/useAuth");
vi.mock("../../services/triviaService");

describe("Home", () => {
  afterEach(() => {
    cleanup();
  });

  const userDataMock = {
    name: "testUser",
    username: "testUsername",
    id: "testId",
    avatar: "testAvatar.png",
  };

  it("should match the snapshot when userData is present and has pending trivia", async () => {
    (useAuth as Mock).mockReturnValue({
      userData: userDataMock,
    });

    (TriviaService.getPendingTrivia as Mock).mockResolvedValue({
      data: {
        pendingTrivia: [
          {
            trivia: {
              _id: "trivia1",
              category: "1",
              difficulty: "easy",
              users: [
                {
                  _id: "testId",
                  username: "testUsername",
                  avatar: "testAvatar.png",
                },
              ],
            },
            completed: false,
          },
        ],
      },
    });

    const { container, getByText } = render(
      <MemoryRouter>
        <Home />
      </MemoryRouter>
    );

    await waitFor(() => {
      const heading = getByText(/Pending trivias/);
      expect(heading).toBeTruthy();
    });

    expect(container).toMatchSnapshot();
  });

  it("should match the snapshot when userData is absent", () => {
    (useAuth as Mock).mockReturnValue({
      userData: null,
    });

    const { container } = render(
      <MemoryRouter>
        <Home />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
