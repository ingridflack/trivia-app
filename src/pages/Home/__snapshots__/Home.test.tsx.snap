// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home > should match the snapshot when userData is absent 1`] = `
<div>
  <header
    class="css-cyjglw"
  >
    <div
      class="chakra-container css-old1by"
    >
      <div
        class="chakra-stack css-2jbh02"
      >
        <a
          href="/"
        >
          <p
            class="chakra-text css-kkaoep"
          >
            Trivia
          </p>
        </a>
        <div
          class="css-j0ozid"
        >
          <a
            href="/login"
          >
             Sign in 
          </a>
          <a
            href="/sign-up"
          >
             Sign Up 
          </a>
        </div>
      </div>
    </div>
  </header>
  <div
    class="css-hkf5rz"
  >
    <div
      class="chakra-container css-1snqisn"
    >
      <p
        class="chakra-text css-141jhoy"
      >
        Challenge yourself in a trivia match!
      </p>
      <a
        class="chakra-button css-4cakko"
        href="/trivia/config"
      >
        Start a trivia
      </a>
      <img
        alt="Home image"
        class="chakra-image css-1tbuem2"
        src="/src/assets/home.svg"
      />
    </div>
  </div>
  <section
    class="css-og67dw"
  >
    <header
      class="css-1fos7ah"
    >
      <h2
        class="chakra-heading css-z7ul1s"
      >
        Show off your knowledge
      </h2>
      <p
        class="chakra-text css-qym2lc"
      >
        Take a quiz and impress your friends with what you know!
      </p>
    </header>
    <div
      class="chakra-container css-14z87jx"
    >
      <div
        class="chakra-card css-1ws3iy8"
      >
        <div
          class="css-q6t6sw"
        >
          <img
            alt="General Knowledge"
            class="chakra-image css-0"
            src="/src/assets/general-knowledge.svg"
          />
        </div>
        <p
          class="chakra-text css-1ey7oos"
        >
          General Knowledge
        </p>
        <button
          class="chakra-button css-1k532n6"
          type="button"
        >
          Play
        </button>
      </div>
      <div
        class="chakra-card css-1ws3iy8"
      >
        <div
          class="css-zuybwl"
        >
          <img
            alt="Books"
            class="chakra-image css-0"
            src="/src/assets/books.svg"
          />
        </div>
        <p
          class="chakra-text css-1ey7oos"
        >
          Books
        </p>
        <button
          class="chakra-button css-1k532n6"
          type="button"
        >
          Play
        </button>
      </div>
      <div
        class="chakra-card css-1ws3iy8"
      >
        <div
          class="css-1gu8hpm"
        >
          <img
            alt="Television"
            class="chakra-image css-0"
            src="/src/assets/tv.svg"
          />
        </div>
        <p
          class="chakra-text css-1ey7oos"
        >
          Television
        </p>
        <button
          class="chakra-button css-1k532n6"
          type="button"
        >
          Play
        </button>
      </div>
      <div
        class="chakra-card css-1ws3iy8"
      >
        <div
          class="css-mtd76j"
        >
          <img
            alt="Mathematics"
            class="chakra-image css-0"
            src="/src/assets/math.svg"
          />
        </div>
        <p
          class="chakra-text css-1ey7oos"
        >
          Mathematics
        </p>
        <button
          class="chakra-button css-1k532n6"
          type="button"
        >
          Play
        </button>
      </div>
      <div
        class="chakra-card css-1ws3iy8"
      >
        <div
          class="css-1fsmisw"
        >
          <img
            alt="Japanese Anime & Manga"
            class="chakra-image css-0"
            src="/src/assets/japanese.svg"
          />
        </div>
        <p
          class="chakra-text css-1ey7oos"
        >
          Japanese Anime & Manga
        </p>
        <button
          class="chakra-button css-1k532n6"
          type="button"
        >
          Play
        </button>
      </div>
    </div>
  </section>
  <footer
    class="css-1s2lyti"
  >
    <div
      class="chakra-container css-0"
    >
      <div
        class="chakra-stack css-11h4xmw"
      >
        <a
          href="/"
        >
          <p
            class="chakra-text css-rekzoj"
          >
            Trivia
          </p>
        </a>
        <div
          class="css-0"
        >
          <strong
            class="chakra-heading css-kf3i1p"
          >
            Site map
          </strong>
          <div
            class="chakra-stack css-1ko7mmx"
          >
            <a
              href="/"
            >
              Home
            </a>
            <a
              href="/login"
            >
              Sign in
            </a>
            <a
              href="/sign-up"
            >
              Sign up
            </a>
          </div>
        </div>
        <div
          class="css-0"
        >
          <strong
            class="chakra-heading css-kf3i1p"
          >
            Social
          </strong>
          <div
            class="chakra-stack css-1ko7mmx"
          >
            <a
              href="https://www.linkedin.com/in/ingrid-flack/"
            >
              LinkedIn
            </a>
            <a
              href="https://github.com/ingridflack"
            >
              GitHub
            </a>
          </div>
        </div>
      </div>
      <div
        class="chakra-stack css-1y3yl0q"
      >
        <p
          class="chakra-text css-0"
        >
          Â© 2024 Trivia App
        </p>
        <p
          class="chakra-text css-0"
        >
          Developed by Ingrid Flack
        </p>
      </div>
    </div>
  </footer>
</div>
`;

exports[`Home > should match the snapshot when userData is present and has pending trivia 1`] = `
<div>
  <header
    class="css-cyjglw"
  >
    <div
      class="chakra-container css-old1by"
    >
      <div
        class="chakra-stack css-2jbh02"
      >
        <a
          href="/"
        >
          <p
            class="chakra-text css-kkaoep"
          >
            Trivia
          </p>
        </a>
        <div
          class="css-j0ozid"
        >
          <a
            href="/trivia/history"
          >
             History 
          </a>
          <a
            href="/login"
          >
            Logout
          </a>
          <span
            class="chakra-avatar css-1et69xt"
          >
            <svg
              aria-label=" avatar"
              class="chakra-avatar__svg css-16ite8i"
              role="img"
              viewBox="0 0 128 128"
            >
              <path
                d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
                fill="currentColor"
              />
              <path
                d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
                fill="currentColor"
              />
            </svg>
          </span>
          <p
            class="chakra-text css-11nn4uy"
          >
             
            testUsername
          </p>
        </div>
      </div>
    </div>
  </header>
  <div
    class="css-hkf5rz"
  >
    <div
      class="chakra-container css-1snqisn"
    >
      <p
        class="chakra-text css-141jhoy"
      >
        Challenge yourself in a trivia match!
      </p>
      <a
        class="chakra-button css-4cakko"
        href="/trivia/config"
      >
        Start a trivia
      </a>
      <img
        alt="Home image"
        class="chakra-image css-1tbuem2"
        src="/src/assets/home.svg"
      />
    </div>
  </div>
  <div
    class="chakra-container css-1clgay9"
  >
    <header
      class="css-8tu9to"
    >
      <h2
        class="chakra-heading css-1tapsqs"
      >
        Pending trivias
      </h2>
    </header>
    <div
      class="chakra-stack css-1sa45dc"
    >
      <div
        class="chakra-card css-2dlz7"
      >
        <span
          class="chakra-badge css-8k89l4"
        >
          easy
        </span>
        <div
          class="chakra-stack css-xzl3c"
        >
          <div
            class="css-7d8kcm"
          >
            <img
              class="chakra-image css-0"
            />
          </div>
          <div
            class="css-0"
          >
            <p
              class="chakra-text css-8ler1h"
            />
          </div>
        </div>
        <a
          class="chakra-button css-10idvi4"
          href="/trivia/trivia1"
        >
          Join
          <div
            class="chakra-avatar__group css-1monvpv"
            role="group"
          >
            <span
              class="chakra-avatar css-r8mluz"
            >
              <div
                aria-label="testUsername"
                class="chakra-avatar__initials css-0"
                role="img"
              >
                t
              </div>
            </span>
          </div>
        </a>
      </div>
    </div>
  </div>
  <section
    class="css-og67dw"
  >
    <header
      class="css-1fos7ah"
    >
      <h2
        class="chakra-heading css-z7ul1s"
      >
        Show off your knowledge
      </h2>
      <p
        class="chakra-text css-qym2lc"
      >
        Take a quiz and impress your friends with what you know!
      </p>
    </header>
    <div
      class="chakra-container css-14z87jx"
    >
      <div
        class="chakra-card css-1ws3iy8"
      >
        <div
          class="css-q6t6sw"
        >
          <img
            alt="General Knowledge"
            class="chakra-image css-0"
            src="/src/assets/general-knowledge.svg"
          />
        </div>
        <p
          class="chakra-text css-1ey7oos"
        >
          General Knowledge
        </p>
        <button
          class="chakra-button css-1k532n6"
          type="button"
        >
          Play
        </button>
      </div>
      <div
        class="chakra-card css-1ws3iy8"
      >
        <div
          class="css-zuybwl"
        >
          <img
            alt="Books"
            class="chakra-image css-0"
            src="/src/assets/books.svg"
          />
        </div>
        <p
          class="chakra-text css-1ey7oos"
        >
          Books
        </p>
        <button
          class="chakra-button css-1k532n6"
          type="button"
        >
          Play
        </button>
      </div>
      <div
        class="chakra-card css-1ws3iy8"
      >
        <div
          class="css-1gu8hpm"
        >
          <img
            alt="Television"
            class="chakra-image css-0"
            src="/src/assets/tv.svg"
          />
        </div>
        <p
          class="chakra-text css-1ey7oos"
        >
          Television
        </p>
        <button
          class="chakra-button css-1k532n6"
          type="button"
        >
          Play
        </button>
      </div>
      <div
        class="chakra-card css-1ws3iy8"
      >
        <div
          class="css-mtd76j"
        >
          <img
            alt="Mathematics"
            class="chakra-image css-0"
            src="/src/assets/math.svg"
          />
        </div>
        <p
          class="chakra-text css-1ey7oos"
        >
          Mathematics
        </p>
        <button
          class="chakra-button css-1k532n6"
          type="button"
        >
          Play
        </button>
      </div>
      <div
        class="chakra-card css-1ws3iy8"
      >
        <div
          class="css-1fsmisw"
        >
          <img
            alt="Japanese Anime & Manga"
            class="chakra-image css-0"
            src="/src/assets/japanese.svg"
          />
        </div>
        <p
          class="chakra-text css-1ey7oos"
        >
          Japanese Anime & Manga
        </p>
        <button
          class="chakra-button css-1k532n6"
          type="button"
        >
          Play
        </button>
      </div>
    </div>
  </section>
  <footer
    class="css-1s2lyti"
  >
    <div
      class="chakra-container css-0"
    >
      <div
        class="chakra-stack css-11h4xmw"
      >
        <a
          href="/"
        >
          <p
            class="chakra-text css-rekzoj"
          >
            Trivia
          </p>
        </a>
        <div
          class="css-0"
        >
          <strong
            class="chakra-heading css-kf3i1p"
          >
            Site map
          </strong>
          <div
            class="chakra-stack css-1ko7mmx"
          >
            <a
              href="/"
            >
              Home
            </a>
            <a
              href="/login"
            >
              Sign in
            </a>
            <a
              href="/sign-up"
            >
              Sign up
            </a>
          </div>
        </div>
        <div
          class="css-0"
        >
          <strong
            class="chakra-heading css-kf3i1p"
          >
            Social
          </strong>
          <div
            class="chakra-stack css-1ko7mmx"
          >
            <a
              href="https://www.linkedin.com/in/ingrid-flack/"
            >
              LinkedIn
            </a>
            <a
              href="https://github.com/ingridflack"
            >
              GitHub
            </a>
          </div>
        </div>
      </div>
      <div
        class="chakra-stack css-1y3yl0q"
      >
        <p
          class="chakra-text css-0"
        >
          Â© 2024 Trivia App
        </p>
        <p
          class="chakra-text css-0"
        >
          Developed by Ingrid Flack
        </p>
      </div>
    </div>
  </footer>
</div>
`;
